pkg/transformers/resolver.go

```diff
@@ -0,0 +1,595 @@
+		case *ast.ImportSpec:
+			// initially, mark all imports as needed
+			key := ""
+			if x.Name != nil {
+				key = x.Name.Name
+			} else {
+				key = x.Path.Value
+			}
+			importsUsed[key] = importUsage{
+				path: strings.Trim(x.Path.Value, `"`),
+				used: true,
+			}
+
+			// keep a hold of the `ResolveReferences` function body so that we can
+		// properly inject variable declarations, error checks, etc. into the
+		// correct positions.
+		case *ast.FuncDecl:
+			block = x.Body
+
+			// keep a hold of the `APIResolver.Resolve` and
+		// `APIResolver.ResolveMultiple` return value assignments as we will
+		// inject code right above it.
+		case *ast.AssignStmt:
+			assign = x
+
+		// we will attempt to transform expressions such as
+		// `reference.To{List: &v1beta1.MRList{}, Managed: &v1beta1.MR{}}`
+		// into:
+		// `reference.To{List: l, Managed: m}`, where
+		// l and m are local variables holding the correctly types MR kind
+		// and MR list kind objects as the reference targets.
+		// Such expressions are the primary sources of cross API-group
+		// import statements.
+		// Cross API-group extractors are rare, and they should be
+		// handled when they're being added, this transformer does not
+		// consider them.
+		case *ast.KeyValueExpr:
+			// check if the key is "To" and the value is a CompositeLit
+			if key, ok := x.Key.(*ast.Ident); ok && key.Name == "To" {
+				// prevent a previous GVK from being reused
+				group, version, kind, listKind = "", "", "", ""
+				if cl, ok := x.Value.(*ast.CompositeLit); ok {
+					// check if there are any package qualifiers in the CompositeLit
+					for _, elt := range cl.Elts {
+						if kv, ok := elt.(*ast.KeyValueExpr); ok {
+							if uexpr, ok := kv.Value.(*ast.UnaryExpr); ok {
+								if cl, ok := uexpr.X.(*ast.CompositeLit); ok {
+									// then the reference target resides in another API group
+									// and the composite literal is a selector expression such as
+									// v1beta1.MR. In this case, we deduce the GV of the MR and
+									// list using the selector expression and the corresponding
+									// import statements (with the name as the expression).
+									// Kind and list kind are determined from the field selector.
+									key := kv.Key.(*ast.Ident).Name
+									if sexpr, ok := cl.Type.(*ast.SelectorExpr); ok {
+										if ident, ok := sexpr.X.(*ast.Ident); ok {
+											path := importsUsed[ident.Name].path
+											importsUsed[ident.Name] = importUsage{
+												path: path,
+												used: false,
+											}
+											// we will parse the import path such as:
+											// github.com/upbound/provider-aws/apis/ec2/v1beta1
+											// and extract the GV information from it.
+											tokens := strings.Split(path, "/")
+											// e.g., v1beta1
+											version = tokens[len(tokens)-1]
+											// e.g., ec2.aws.upbound.io
+											group = fmt.Sprintf("%s.%s", tokens[len(tokens)-2], apiGroupSuffix)
```

Comment:
Will this statement cause any issues with the Azure-ResourceGroup resource? Because the API group of ResourceGroup is the same as the `apiGroupSuffix.: `azure.upbound.io`. 

I mean, the split statement will be this: `github.com/upbound/provider-azure/apis/azure/v1beta1`

Then, the calculated group will be `azure.azure.upbound.io`.

Reply:
Thanks @sergenyalcin, good catch indeed! Given this a try with `upbound/provider-azure` as you suggested and what we generate is:
```go
m, l, err = apisresolver.GetManagedResource("azure.azure.upbound.io", "v1beta1", "ResourceGroup", "ResourceGroupList")
		if err != nil {
			return errors.Wrap(err, "failed to get the reference target managed resource and its list for reference resolution")
		}
```

Added the `--api-group-override` command-line option (shorthand `-o`) so that any API group overrides can be configured. Also added the `--api-resolver-package` (shorthand `-a`) command-line option so that we can properly configure the API resolver function's (`GetManagedResource`) package.

So, when we run the transformer with the following command-line options:
```shell
go run github.com/crossplane/upjet/cmd/resolver -g azure.upbound.io -a github.com/upbound/provider-azure/internal/apis -o azure.azure.upbound.io=azure.upbound.io
```
what gets now generated is:
```go
		m, l, err = apisresolver.GetManagedResource("azure.upbound.io", "v1beta1", "ResourceGroup", "ResourceGroupList")
		if err != nil {
			return errors.Wrap(err, "failed to get the reference target managed resource and its list for reference resolution")
		}
```

In a future iteration (especially if some other providers also adopt this transformer), we may also generate the `GetManagedResource` function via upjet for the provider and remove the `--api-resolver-package` option.